README.md
# [JProbLog](https://julianmendez.github.io/jproblog/)

[![build](https://travis-ci.org/julianmendez/jproblog.png?branch=master)](https://travis-ci.org/julianmendez/jproblog)
[![maven central](https://maven-badges.herokuapp.com/maven-central/de.tu-dresden.inf.lat.jproblog/jproblog/badge.svg)](https://search.maven.org/#search|ga|1|g%3A%22de.tu-dresden.inf.lat.jproblog%22)
[![license](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://www.apache.org/licenses/LICENSE-2.0.txt)


**JProbLog** is a [Java](https://www.oracle.com/technetwork/java/index.html) adapter for [ProbLog](https://dtai.cs.kuleuven.be/problog/).


## Download

* [library JAR file](https://sourceforge.net/projects/latitude/files/jproblog/0.1.0/jproblog-0.1.0.jar/download)
* [The Central Repository](https://repo1.maven.org/maven2/de/tu-dresden/inf/lat/jproblog/)
* as dependency:

```xml
<dependency>
  <groupId>de.tu-dresden.inf.lat.jproblog</groupId>
  <artifactId>jproblog</artifactId>
  <version>0.1.0</version>
</dependency>
```


## Author

[Julian Mendez](https://julianmendez.github.io)


## License

This software is distributed under the [Apache License Version 2.0](https://www.apache.org/licenses/LICENSE-2.0.txt).


## Release notes

See [release notes](https://julianmendez.github.io/jproblog/RELEASE-NOTES.html).


## Source code

To checkout and compile the project, use:

```
$ git clone https://github.com/julianmendez/jproblog.git
$ cd jproblog
$ mvn clean install
```

To compile the project offline, first download the dependencies:

```
$ mvn dependency:go-offline
```

and once offline, use:

```
$ mvn --offline clean install
```

The bundles uploaded to [Sonatype](https://oss.sonatype.org/) are created with:

```
$ mvn clean install -DperformRelease=true
```

and then:

```
$ cd jproblog/target
$ jar -cf bundle.jar jproblog-*
```

The version number is updated with:

```
$ mvn versions:set -DnewVersion=NEW_VERSION
```

where *NEW_VERSION* is the new version.


## Contact

In case you need more information, please contact @julianmendez .






RELEASE-NOTES.md



## Release notes

| version | release date | Java | ProbLog | Python     | Jython  |
|:--------|:-------------|:----:|:-------:|:----------:|:-------:|
| v0.1.0  | 2016-09-01   | 8    | 2.1     | 2.7 or 3.2 | 2.7.0   |


### v0.1.0
*(2016-09-01)*
* requires Java 8
* downloads ProbLog 2.1
* if Python 2.7 or 3.2 is not installed, it downloads Jython 2.7.0 from the Central Repository, and it uses Jython instead
* build commands:
```
$ mvn clean install
```
* release: `jproblog/target/jproblog-0.1.0.jar`




jproblog/src/main/java/de/tudresden/inf/lat/jproblog/JProblog.java
package de.tudresden.inf.lat.jproblog;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Reader;
import java.io.UncheckedIOException;
import java.io.Writer;
import java.util.Objects;
import java.util.function.Function;

/**
 * An object of this class is a Java adapter to execute ProbLog.
 * 
 * @author Julian Mendez
 *
 */
public class JProblog implements Function<String, String> {

	static final char NEW_LINE = '\n';

	private final ProblogProcessor processor;

	public static final String HELP = "Usage: java -jar jproblog.jar <input_file> <output_file>\n\n";

	/**
	 * Constructs a new JProblog.
	 */
	public JProblog() {
		this.processor = new ProblogProcessor();
		this.processor.startInstallation();
	}

	/**
	 * Returns the result of running ProbLog with the given input.
	 * 
	 * @param input
	 *            input
	 * @return the result of running ProbLog with the given input
	 */
	@Override
	public String apply(String input) {
		Objects.requireNonNull(input);
		return this.processor.apply(input);
	}

	String read(BufferedReader reader) {
		StringBuilder sb = new StringBuilder();
		reader.lines().forEach(line -> {
			sb.append(line);
			sb.append(NEW_LINE);
		});
		return sb.toString();
	}

	void write(BufferedWriter writer, String content) throws IOException {
		writer.write(content);
		writer.flush();
	}

	/**
	 * Reads the input from the reader, executes ProbLog, and writes the output
	 * to the writer.
	 * 
	 * @param reader
	 *            reader
	 * @param writer
	 *            writer
	 */
	public void apply(Reader reader, Writer writer) {
		Objects.requireNonNull(reader);
		Objects.requireNonNull(writer);
		try {
			String input = read(new BufferedReader(reader));
			String output = apply(input);
			write(new BufferedWriter(writer), output);
		} catch (IOException e) {
			throw new UncheckedIOException(e);
		}
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		} else if (!(obj instanceof JProblog)) {
			return false;
		} else {
			JProblog other = (JProblog) obj;
			return this.getProcessor().equals(other.getProcessor());
		}
	}

	@Override
	public int hashCode() {
		return this.processor.hashCode();
	}

	@Override
	public String toString() {
		return this.processor.toString();
	}

	/**
	 * Returns the processor used to execute ProbLog.
	 * 
	 * @return the processor used to execute ProbLog
	 */
	public ProblogProcessor getProcessor() {
		return this.processor;
	}

	/**
	 * This is the main method that is executed from the command line.
	 * 
	 * @param args
	 *            command line arguments
	 */
	public static void main(String[] args) {
		JProblog instance = new JProblog();
		if (args.length == 2) {
			try {
				instance.apply(new FileReader(args[0]), new FileWriter(args[1]));
			} catch (IOException e) {
				throw new UncheckedIOException(e);
			}
		} else {
			System.out.println(HELP);
		}
	}

}


jproblog/src/main/java/de/tudresden/inf/lat/jproblog/ProblogProcessor.java
package de.tudresden.inf.lat.jproblog;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.UncheckedIOException;
import java.net.URI;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.IntStream;
import java.util.zip.ZipEntry;

/**
 * An object of this class manages the installation and execution of ProbLog.
 * The processor detects whether a valid version of Python is already installed.
 * If Python is installed, the processor is in 'Python mode' and it uses Python
 * to run ProbLog. Otherwise, it downloads Jython from the Central Repository,
 * and uses Jython to run ProbLog.
 * 
 * @author Julian Mendez
 *
 */
public class ProblogProcessor implements Function<String, String> {

	public static final String JYTHON_GROUP_ID = "org.python";
	public static final String JYTHON_ARTIFACT_ID = "jython-standalone";
	public static final String JYTHON_VERSION = "2.7.0";

	public static final String MAVEN_REPOSITORY = "https://repo1.maven.org/maven2";

	static final char SPACE_CHAR = ' ';
	static final char NEW_LINE_CHAR = '\n';
	static final String TAB_AND_COLON = "\t    : ";
	static final char HYPHEN_CHAR = '-';
	static final char SLASH_CHAR = '/';

	public static final String JYTHON_JAR = "jython-standalone" + HYPHEN_CHAR + JYTHON_VERSION + ".jar";

	public static final URI DEFAULT_JYTHON_DOWNLOAD_URI = URI
			.create(MAVEN_REPOSITORY + SLASH_CHAR + JYTHON_GROUP_ID.replace('.', SLASH_CHAR) + SLASH_CHAR
					+ JYTHON_ARTIFACT_ID + SLASH_CHAR + JYTHON_VERSION + SLASH_CHAR + JYTHON_JAR);

	public static final URI DEFAULT_PROBLOG_DOWNLOAD_URI = URI
			.create("https://bitbucket.org/problog/problog/get/master.zip");

	static final String FILE_SEPARATOR = System.getProperty("file.separator");

	static final String USER_HOME_DIRECTORY = System.getProperty("user.home");

	static final String JPROBLOG_WORKING_DIRECTORY = USER_HOME_DIRECTORY + FILE_SEPARATOR + ".cache" + FILE_SEPARATOR
			+ "jproblog";

	static final String DEFAULT_INPUT_FILE_FOR_PROBLOG = JPROBLOG_WORKING_DIRECTORY + FILE_SEPARATOR
			+ "input_for_problog.txt";

	static final String DEFAULT_OUTPUT_FILE_FROM_PROBLOG = JPROBLOG_WORKING_DIRECTORY + FILE_SEPARATOR
			+ "output_from_problog.txt";

	static final String DEFAULT_PROBLOG_INSTALLATION_DIRECTORY = JPROBLOG_WORKING_DIRECTORY;

	static final String DEFAULT_PROBLOG_ZIP_FILE = JPROBLOG_WORKING_DIRECTORY + FILE_SEPARATOR
			+ "problog-2.1-SNAPSHOT.zip";

	static final String DEFAULT_JYTHON_FILE = JPROBLOG_WORKING_DIRECTORY + FILE_SEPARATOR + JYTHON_JAR;

	static final String PROBLOG_CLI = "problog-cli.py";
	static final String PROBLOG_INSTALL_COMMAND = "install";
	static final String PROBLOG_OUTPUT_OPTION = "-o";

	static final String PYTHON_COMMAND = "python";
	static final String PYTHON_VERSION_OPTION = "-V";
	static final String PYTHON_VERSION_2_7 = "Python 2.7";
	static final String PYTHON_VERSION_3_2 = "Python 3.2";
	static final String JAVA_COMMAND = "java";
	static final String JAVA_JAR_OPTION = "-jar";
	static final String JYTHON_COMMAND = JAVA_COMMAND + SPACE_CHAR + JAVA_JAR_OPTION + SPACE_CHAR + DEFAULT_JYTHON_FILE;

	static final String PROBLOG_EXEC_WINDOWS = "problog" + FILE_SEPARATOR + "bin" + FILE_SEPARATOR + "windows"
			+ FILE_SEPARATOR + "dsharp.exe";
	static final String PROBLOG_EXEC_DARWIN = "problog" + FILE_SEPARATOR + "bin" + FILE_SEPARATOR + "darwin"
			+ FILE_SEPARATOR + "dsharp";
	static final String PROBLOG_EXEC_LINUX = "problog" + FILE_SEPARATOR + "bin" + FILE_SEPARATOR + "linux"
			+ FILE_SEPARATOR + "dsharp";

	public static final String NO_RESULTS = "No results.";

	public static final int KNOWN_ISSUE_MAX_RETRIES = 5;
	public static final String KNOWN_ISSUE_MESSAGE = "DSharpError: DSharp has encountered an error. This is a known issue. See KNOWN_ISSUES for details on how to resolve this problem.";

	private String problogDirectory = null;
	private Object problogInstallationMonitor = new Object();
	private final boolean pythonMode;

	/**
	 * Constructs a new ProbLog processor
	 */
	public ProblogProcessor() {
		this.pythonMode = isPythonInstalled();
	}

	ProblogProcessor(boolean pythonMode) {
		this.pythonMode = pythonMode;
	}

	boolean isPythonInstalled() {
		try {
			ProcessBuilder builder = new ProcessBuilder(PYTHON_COMMAND, PYTHON_VERSION_OPTION);
			Process process = builder.start();
			process.waitFor();
			BufferedReader reader = new BufferedReader(new InputStreamReader(process.getErrorStream()));
			String line = reader.readLine();
			return (line != null && (line.startsWith(PYTHON_VERSION_2_7) || line.startsWith(PYTHON_VERSION_3_2)));
		} catch (IOException e) {
			throw new UncheckedIOException(e);
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}

	/**
	 * Tells whether this processor is in 'Python mode', i.e. it uses Python
	 * instead of Jython.
	 * 
	 * @return <code>true</code> if this processor is in Python mode
	 */
	public boolean getPythonMode() {
		return this.pythonMode;
	}

	String flattenArguments(String[] args) {
		StringBuilder sb = new StringBuilder();
		IntStream.range(0, args.length).forEach(index -> {
			sb.append(SPACE_CHAR);
			sb.append(args[index]);
		});
		return sb.toString();
	}

	int runPython(String[] args) {
		try {
			String commandLine = PYTHON_COMMAND + flattenArguments(args);
			Runtime runtime = Runtime.getRuntime();
			Process process = runtime.exec(commandLine);
			return process.waitFor();
		} catch (IOException e) {
			throw new UncheckedIOException(e);
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}

	int runJython(String[] args) {
		try {
			String commandLine = JYTHON_COMMAND + flattenArguments(args);
			Runtime runtime = Runtime.getRuntime();
			Process process = runtime.exec(commandLine);
			return process.waitFor();
		} catch (IOException e) {
			throw new UncheckedIOException(e);
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}

	int run(String args[]) {
		if (this.pythonMode) {
			return runPython(args);
		} else {
			return runJython(args);
		}
	}

	/**
	 * Decompresses the ProbLog ZIP file and returns the directory name.
	 * 
	 * @param problogZipFile
	 *            file name of ProbLog ZIP file
	 * @param problogDirectory
	 *            directory where ProbLog is being installed
	 * @return the directory name
	 * @throws IOException
	 *             if something goes wrong with I/O
	 */
	String decompressProblog(String problogZipFile, String problogDirectory) throws IOException {
		Objects.requireNonNull(problogZipFile);
		Objects.requireNonNull(problogDirectory);
		Decompressor installer = new Decompressor();
		ZipEntry directory = installer.decompress(new File(problogZipFile), new File(problogDirectory));
		return directory.getName();
	}

	/**
	 * Updates execute permission of key executable files.
	 * 
	 * @param problogDirectory
	 *            directory where ProbLog is being installed
	 * @throws IOException
	 *             if something goes wrong with I/O
	 */
	void updatePermissions(String problogDirectory) throws IOException {
		Objects.requireNonNull(problogDirectory);
		(new File(problogDirectory + FILE_SEPARATOR + PROBLOG_EXEC_WINDOWS)).setExecutable(true);
		(new File(problogDirectory + FILE_SEPARATOR + PROBLOG_EXEC_DARWIN)).setExecutable(true);
		(new File(problogDirectory + FILE_SEPARATOR + PROBLOG_EXEC_LINUX)).setExecutable(true);
	}

	/**
	 * Installs ProbLog. This is necessary because just decompressing the ZIP
	 * file is not enough.
	 * 
	 * @param problogDirectory
	 *            directory where ProbLog has been installed
	 * @return the exit value given by the operating system
	 * @throws IOException
	 *             if something goes wrong with I/O
	 * @throws InterruptedException
	 *             if the execution was interrupted
	 */
	int installProblog(String problogDirectory) throws IOException, InterruptedException {
		String[] args = new String[2];
		args[0] = problogDirectory + FILE_SEPARATOR + PROBLOG_CLI;
		args[1] = PROBLOG_INSTALL_COMMAND;
		return run(args);
	}

	public String getProblogDirectory() {
		return this.problogDirectory;
	}

	/**
	 * Downloads and installs ProbLog.
	 * 
	 * @throws IOException
	 *             if something goes wrong with I/O
	 * @throws InterruptedException
	 *             if the execution was interrupted
	 */
	public void install() throws IOException, InterruptedException {
		if (!this.pythonMode) {
			DownloadManager jythonDownloadManager = new DownloadManager(DEFAULT_JYTHON_DOWNLOAD_URI,
					DEFAULT_JYTHON_FILE);
			jythonDownloadManager.downloadIfNecessary();
		}

		DownloadManager problogDownloadManager = new DownloadManager(DEFAULT_PROBLOG_DOWNLOAD_URI,
				DEFAULT_PROBLOG_ZIP_FILE);
		problogDownloadManager.downloadIfNecessary();

		String directory = decompressProblog(problogDownloadManager.getFileName(),
				DEFAULT_PROBLOG_INSTALLATION_DIRECTORY);
		this.problogDirectory = DEFAULT_PROBLOG_INSTALLATION_DIRECTORY + FILE_SEPARATOR + directory;
		updatePermissions(problogDirectory);
		installProblog(problogDirectory);
	}

	/**
	 * Starts the installation of ProbLog.
	 */
	public void startInstallation() {
		Thread thread = new Thread() {
			public void run() {
				synchronized (problogInstallationMonitor) {
					try {
						install();
					} catch (IOException | InterruptedException e) {
						throw new RuntimeException(e);
					}
				}
			}
		};
		thread.start();
	}

	/**
	 * Executes ProbLog and returns the exit value given by the operating
	 * system.
	 * 
	 * @param outputFileName
	 *            file name of output
	 * @return the exit value given by the operating system
	 */
	public int execute(String outputFileName) {
		Objects.requireNonNull(outputFileName);
		synchronized (this.problogInstallationMonitor) {
			String[] args = new String[4];
			args[0] = this.problogDirectory + FILE_SEPARATOR + PROBLOG_CLI;
			args[1] = (new File(DEFAULT_INPUT_FILE_FOR_PROBLOG)).getAbsolutePath();
			args[2] = PROBLOG_OUTPUT_OPTION;
			args[3] = (new File(outputFileName)).getAbsolutePath();
			return run(args);
		}
	}

	/**
	 * Returns string with the content of a given reader.
	 * 
	 * @param input
	 *            reader
	 * @return string with the content of a given reader
	 */
	String show(Reader input) {
		Objects.requireNonNull(input);
		StringBuilder sb = new StringBuilder();
		BufferedReader reader = new BufferedReader(input);
		reader.lines().forEach(line -> {
			sb.append(line);
			sb.append(NEW_LINE_CHAR);
		});
		return sb.toString();
	}

	void createInputFileForProblog(String input, String inputFileForProblog) throws IOException {
		BufferedWriter writer = new BufferedWriter(new FileWriter(inputFileForProblog));
		writer.write(input);
		writer.flush();
		writer.close();
	}

	String tryExecution() throws IOException {
		File outputFile = new File(DEFAULT_OUTPUT_FILE_FROM_PROBLOG);
		outputFile.delete();
		execute(DEFAULT_OUTPUT_FILE_FROM_PROBLOG);
		if (outputFile.exists()) {
			return show(new InputStreamReader(new FileInputStream(outputFile)));
		} else {
			return NO_RESULTS;
		}
	}

	@Override
	public String apply(String input) {
		Objects.requireNonNull(input);
		try {
			createInputFileForProblog(input, DEFAULT_INPUT_FILE_FOR_PROBLOG);
			String ret = "";
			for (int iteration = 0; iteration < KNOWN_ISSUE_MAX_RETRIES
					&& (ret.isEmpty() || ret.trim().equals(KNOWN_ISSUE_MESSAGE)); iteration++) {
				ret = tryExecution();
			}
			return ret;
		} catch (IOException e) {
			throw new UncheckedIOException(e);
		}
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		} else if (!(obj instanceof ProblogProcessor)) {
			return false;
		} else {
			ProblogProcessor other = (ProblogProcessor) obj;
			return this.getProblogDirectory().equals(other.getProblogDirectory());
		}
	}

	@Override
	public int hashCode() {
		return toString().hashCode();
	}

	@Override
	public String toString() {
		return this.problogDirectory;
	}

}
